### PR TITLE
[Merged by Bors] - feat: Better lemmas for transferring finite sums along equivalences

### PR DESCRIPTION
**Pull request by YaelDillies**
Lemmas around this were a mess, throth in terms of names, statement and location. This PR standardises everything to be in `Algebra.BigOperators.Basic` and changes the lemmas to take in `InjOn` and `SurjOn` assumptions where possible (and where impossible make sure the hypotheses are taken in the correct order) and moves the equality of functions hypothesis last.

Also add a few lemmas that help fix downstream uses by golfing.

From LeanAPAP and LeanCamCombi

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/BigOperators/Basic.lean [hunk 1]
```diff
@@ -569,47 +572,94 @@ theorem prod_bij {s : Finset α} {t : Finset γ} {f : α → β} {g : γ → β}
...
-      (fun a _ha => by simp_rw [e.apply_symm_apply]) (fun i _hi => e i) (fun a ha => _)
-      (fun a _ha => e.apply_symm_apply a) fun a _ha => e.symm_apply_apply a
-  rcases Finset.mem_image.mp ha with ⟨i', hi', rfl⟩
-  dsimp only
-  rwa [e.apply_symm_apply]
-#align finset.equiv.prod_comp_finset Finset.Equiv.prod_comp_finset
-#align finset.equiv.sum_comp_finset Finset.Equiv.sum_comp_finset
+-- TODO: Backport arguments changes to `card_congr` and `prod_bij`
+@[to_additive]
+lemma prod_nbij (i : ι → κ) (hi : ∀ a ∈ s, i a ∈ t) (i_inj : (s : Set ι).InjOn i)
```
Comment (jcommelin) [inline Mathlib/Algebra/BigOperators/Basic.lean:587]: What does the `n` in `nbij` stand for?

**Comment by jcommelin** [comment]
!bench

File: Mathlib/Algebra/BigOperators/Basic.lean (hunk 1 above)
Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Basic.lean:587]: **N**on-dependent. Bhavik came up with this name, we both don't like it. Do you have any suggestion?

**Comment by YaelDillies** [comment]
!bench

File: Mathlib/Algebra/BigOperators/Basic.lean (hunk 1 above)
Comment (jcommelin) [reply Mathlib/Algebra/BigOperators/Basic.lean:581]: I'm inclined to go with `prod_bij'`. And just explain the difference in the docstrings using cross-refs. "See `foo_bar` for a (non)dependent variant."

Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Basic.lean:581]: `prod_bij'` already exists! It's the version which takes in two functions that are inverse to each other on the finsets.

**Comment by YaelDillies** [comment]
`Probability.StrongLaw` basically proves the same lemma about swapping sums three times and I've fixed everything by sticking in a lot of `aesop` calls. I will try to find a generalisation of `sum_Ico_Ico_comm` that works for all three applications, but I'm willing to call that out of scope for this PR.

File: Mathlib/Algebra/BigOperators/Basic.lean (hunk 1 above)
Comment (jcommelin) [reply Mathlib/Algebra/BigOperators/Basic.lean:581]: Aha... arguably the name of that lemma should be closer to `equiv` than to `bij`. But maybe that's out of scope for this PR.

So let's go for `nbij` for now. But please explain the name in the docstrings.

**Comment by jcommelin** [comment]
bors d+

**Comment by YaelDillies** [comment]
bors merge

**Comment by YaelDillies** [comment]
bors merge

**Pull request closed without merging**